File: Day79/C_Little_Girl_and_Maximum_Sum.py
n,m=list(map(int,input().split()))
a = list(map(int,input().split()))
q = [0]*m
dist = [0] * (n+1)
for i in range(m):
    l,r = list(map(int,input().split()))
    q[i] = (l,r)
    dist[l-1] += 1
    dist[r] -= 1
d = [(dist[0],0)]
for i in range(1,n):
    dist[i] += dist[i-1]
    d.append((dist[i],i))
d.sort(key=lambda item:item[0],reverse=True)
a.sort(reverse=True)
print(d)
print(a)
final = [0]*n

for i in range(n):
    final[d[i][1]] = a[i]
print(final)
prefixSum = [0]
for i in range(n):
    prefixSum.append(prefixSum[-1] + final[i])
result = 0
for j in q:
    result += prefixSum[j[1]] - prefixSum[j[0]-1]
print(result)

[(3,1),(2,2),(2,0),]
[5,3,2,]
[2,5,3,]
9

[(3, 1), (2, 0), (2, 2)]
[5, 3, 2]
[3, 5, 2]
25

File: Day79/C_Little_Girl_and_Maximum_Sum.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

typedef long long LL;

bool sortcol(vector<int>& v1, vector<int>& v2)
{
    return v1[0] < v2[0];
}

int main()
{
    int n, m;
    cin >> n >> m;
    vector<int> a;
    for (int i = 0; i < n; i++) {
        LL x;
        cin >> x;
        a.push_back(x);
    }
    int q[m][2];
    int dist[n + 1];
    std::fill_n(dist, n + 1, 0);
    for (int i = 0; i < m; i++) {
        cin >> q[i][0] >> q[i][1];
        dist[q[i][0]-1] += 1;
        dist[q[i][1]] -= 1;
    }

    vector< vector<int> > d;
    vector<int> v;
    v.push_back(dist[0]);
    v.push_back(0);
    d.push_back(v);

    for (int i = 1; i < n; i++) {
        dist[i] += dist[i - 1];
        vector<int> v;
        v.push_back(dist[i]);
        v.push_back(i);
        d.push_back(v);
    }
    sort(d.begin(), d.end(), sortcol);
    reverse(d.begin(), d.end());
    sort(a.begin(), a.end());
    reverse(a.begin(), a.end());

    int final[n];
    std::fill_n(final, n, 0);

    for (int i = 0; i < n; i++) {
        final[d[i][1]] = a[i];
    }
    vector<LL> prefixSum;
    prefixSum.push_back(0);

    for (int i = 0; i < n; i++) {
        prefixSum.push_back(*prefixSum.rbegin() + final[i]);
    }
    LL result = 0;
    for (int j = 0; j < m; j++) {
        result += prefixSum[q[j][1]] - prefixSum[q[j][0] - 1];
    }
    cout << result << endl;

    return 0;
}